### DISCUSSION TITLE
cli should not require camelCase style

### DISCUSSION THREAD

**Discussion by spaceunifyfifty** [sbt/sbt#7601] [General]
Initially discussed at Scala.js [here](https://github.com/scala-js/scala-js/issues/5006).

Commands such as `sbt testQuick`, `sbt fastLinkJS` are camel-cased, and the lowercased forms `testquick`, `fastlinkjs` are rejected as invalid.

camelCasing is idiomatic for source code, but a bit weird and tiresome for the command-line. As far as I know, it's not common in the unix world (where cli usage is heavy and users want to be able to execute commands effortlessly and fast) to have to worry about uppercasing every subsequent word in a command. (It's small, but gets noticeably more irritating the more you do it). The sbt cli would feel a bit more ergonomic for people of that unix background if lower-cased versions of the command were acceptable, even if just as aliases.

I'm curious what the motivation was in choosing this style over kebab-case like other cli tools, if anyone knows.

**Comment by SethTisue** [comment]
I think sbt was designed primarily with the JVM in mind rather primarily with the Unix command line in mind. We do have `sbtn` (aka `sbt -client`) these days, but that was a very late arrival. For most of sbt's lifetime, and even for many (most?) current users, the primary means of interacting with it was to run `sbt` and then issue any further commands from inside the sbt shell. (And the sbt shell has very little overlap in functionality with the Unix shell.)

also,

> camelCasing is idiomatic for source code

Yes, I believe the sbt design reflects that fact that sbt build definitions _are_ Scala source code. Settings and tasks are `val`s. When you use the shell you're "talking" to some Scala code. And that Scala code isn't just behind-the-scenes stuff; many builds contain custom tasks and settings. And I suspect there is an ambient assumption that the person _using_ the build is the same person who is _writing_ the build, so the fact that Scala code is involved is considered to be core to the design, rather than just an implementation detail.

**Comment by eed3si9n** [comment]
> I'm curious what the motivation was in choosing this style over kebab-case like other cli tools, if anyone knows.

sbt initially did have that design, as you can see in older documentations like https://www.scala-sbt.org/0.12.4/docs/Getting-Started/Basic-Def.html#keys-in-sbt-interactive-mode:

> In build definition files, keys are named with **camelCase** following Scala convention, but the sbt command line uses hyphen-separated-words instead. The **hyphen-separated** string used in sbt comes from the definition of the key (see [Keys](https://www.scala-sbt.org/0.12.4/sxr/Keys.scala.html)). For example, in Keys.scala, there's this key:
>
> ```scala
>  val scalacOptions = TaskKey[Seq[String]]("scalac-options", "Options for the Scala compiler.")
> ```
>
> In sbt you type `scalac-options` but in a build definition file you use `scalacOptions`.

With tab completion, the idea of kebab case sounded like a cool idea, but in practice the idea of having two names for one key was extremely confusing to the end users. Because of scoping we had to something like `scalacOptions in Compile` in `build.sbt` file, but `compile:scalac-options` on the sbt shell or batch mode (OS shell).

During sbt 0.13 and 1.x, Josh (maintainer, second) and I (maintainer, third) have:
1. unify the settings and task keys to match the Scala variable names `scalacOptions`
2. unify the scoping syntax on both `build.sbt` and shell to using [slash](https://contributors.scala-lang.org/t/unification-of-sbt-shell-notation-and-build-sbt-dsl/913)

Overall these changes are considered to have made sbt easier to understand.
 
>  The sbt cli would feel a bit more ergonomic for people of that unix background if lower-cased versions of the command were acceptable, even if just as aliases.

As Seth mentioned, given that we now have sbtn, maybe we should reevaluate the unify-to-build.sbt course, and consider admitting lower-case as well. For a short-term workaround consider defining your own alias:

https://github.com/sbt/sbt/blob/1d4f7257a45847c07f15e01f868e1fd2a7095871/sbt-app/src/sbt-test/actions/aliasrc/.sbtrc#L2
